src/server/game/Entities/GameObject/GameObject.cpp

```diff
@@ -2411,6 +2434,122 @@ void GameObject::SetGoState(GOState state)
         else if (state == GO_STATE_READY)
             EnableCollision(!startOpen);*/
     }
+    /* Whenever a gameobject inside an instance changes
+     * save it's state on the database to be loaded properly
+     * on server restart or crash.
+     */
+    if (IsInstanceGameobject())
+    {
+        // Save the gameobject state on the Database
+        if (!FindStateSavedOnInstance())
+        {
+            SaveInstanceData(GetGameobjectStateAsShort(&state));
+        }
+        else
+        {
+            UpdateInstanceData(GetGameobjectStateAsShort(&state));
+        }
+    }
+}
+
+bool GameObject::IsInstanceGameobject()
+{
+    // Avoid checking for unecessary gameobjects whose
+    // states don't matter for the dungeon progression
+    if (!ValidateGameobjectType())
+    {
+        return false;
+    }
+
+    if (auto* map = FindMap())
+    {
+        if (map->IsDungeon() || map->IsRaid())
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool GameObject::ValidateGameobjectType()
+{
+    switch (m_goInfo->type)
+    {
+        case GAMEOBJECT_TYPE_DOOR:
+        case GAMEOBJECT_TYPE_BUTTON:
+        case GAMEOBJECT_TYPE_TRAP:
+        case GAMEOBJECT_TYPE_DESTRUCTIBLE_BUILDING:
+        case GAMEOBJECT_TYPE_TRAPDOOR:
+            return true;
+        default:
+            return false;
+    }
+}
+
+unsigned short GameObject::GetGameobjectStateAsShort(GOState* state)
+{
+    unsigned short m_state = 3;
```

Comment:
uint16

Reply:
or uint8, just not unsigned short

Reply:
should match the gameobject *state* attribute imo

Reply:
Yea uint8, the DB table should change to TINYINT

Reply:
And this func should be renamed as well

Reply:
(state in `gameobject` is TINYINT)